### ISSUE TITLE
EACCEPT returns SGX_PAGE_ATTRIBUTES_MISMATCH

### ISSUE THREAD

**Issue by alexthomas1** [intel/linux-sgx#948]
If enclave_alloc succeeds with a virtual address in-enclave, what reasons would running EACCEPT on the address fail afterwards? I get an error message when I run eaccept on a virtual address which has been previously alloced via enclave_alloc.

The return code I get matches: SGX_PAGE_ATTRIBUTES_MISMATCH, how do i fix this?

I see that someone has the psuedo code for it here: https://github.com/Mic92/x86_64-linux-cheatsheats/blob/master/pages/ENCLU%5BEACCEPT%5D

How do I know what the correct EPCM page attributes are?

**Comment by lzha101** [comment]
From Intel SDM,  EACCEPT instruction returns SGX_PAGE_ATTRIBUTES_MISMATCH if below check fails: 

> (* Verify that accept request matches current EPC page settings *)
IF ( (EPCM(DS:RCX).ENCLAVEADDRESS ≠ DS:RCX) or (EPCM(DS:RCX).PENDING ≠ SCRATCH_SECINFO.FLAGS.PENDING) or
(EPCM(DS:RCX).MODIFIED ≠ SCRATCH_SECINFO.FLAGS.MODIFIED) or (EPCM(DS:RCX).R ≠ SCRATCH_SECINFO.FLAGS.R) or
(EPCM(DS:RCX).W ≠ SCRATCH_SECINFO.FLAGS.W) or (EPCM(DS:RCX).X ≠ SCRATCH_SECINFO.FLAGS.X) or
(EPCM(DS:RCX).PT ≠ SCRATCH_SECINFO.FLAGS.PT) )
THEN
RFLAGS.ZF <- 1;
RAX  <- SGX_PAGE_ATTRIBUTES_MISMATCH;
GOTO DONE;
FI;

Not sure how you call EACCEPT on the virtual addr. Just notice that in sgx-emm repo, [do_commit](https://github.com/intel/sgx-emm/blob/222b9d8cd246809cc8da041241824b0d0469000f/ema.c#L645) function, it sets the PENDING flag  before call eaccept to commit the memory: 

> sec_info_t si SGX_SECINFO_ALIGN = {si_flags | SGX_EMA_STATE_PENDING, 0};
    int ret = -1;

You could check whether you have done the same thing before trigger EACCEPT.

**Comment by alexthomas1** [comment]
Yes, I tried every single combination of RWX and also set the pending flag as well. I also make sure that the PT is a regular page (but I also tried combinations of TCS/PT_TRIM configurations that were allowed). I am trying to get edmm working for OpenEnclave. The particular page I am trying to eaccept, I call enclave_alloc on it first, then eaccept. I actually just copy the entire EDMM code provided in the sgx-emm repo to achieve this. What can I do to figure out the right attributes to provide? Is the address I pass not correct (I notice that's part of the check EPCM(DS:RCX).ENCLAVEADDRESS ≠ DS:RCX)?

**Comment by haitaohuang** [comment]
if you are using sgx-emm, then you do not need to do eaccept on your own. When sgx_mm_alloc is invoked, you provide either COMMIT_ON_DEMAND for eaccept done on PF or COMMIT_NOW for EACCEPT done during sgx_mm_alloc call.

**Comment by alexthomas1** [comment]
I mean when I call sgx_mm_alloc, the eaccept portion of it fails. I take the existing edmm code and try to call it in an enclave in OpenEnclave (after calling sgx_mm_init with appropriate start and end addresses), but the EACCEPT fails and gives me SGX_PAGE_ATTRIBUTES_MISMATCH

**Comment by haitaohuang** [comment]
I notice the OE edmm branch reserves a hard coded 40 page range at the end of enclave: https://github.com/mingweishih/openenclave/blob/b07dc610c34b4cd3bc050d3b5427177273974a4d/enclave/core/sgx/edmm.c#L166. And it does not mark runtime usage (code/data preloaded with EADD before EINIT)
You need make sure enclave size is big enough, so that user_start(user_end-40*OE_PAGE_SIZE) would not be in those ranges.
Otherwise, emm would try to re-allocate those already used pages and try to EACCEPT which would fail.

**Comment by alexthomas1** [comment]
Oh, there's an EDMM branch for OpenEnclave... I was actually trying to implement this in OpenEnclave. Thank you, I will just use this branch.

**Comment by haitaohuang** [comment]
Check with @mingweishih

**Comment by mingweishih** [comment]
The branch was the poc that I made for the initial version of emm. It might need some tweaks (or not) to make it work with current version of emm. I will recommend you just use it as a reference.

**Comment by alexthomas1** [comment]
Can you explain why the hardcoded range of user_end-40*OE_PAGE_SIZE to user_end works? I tried setting it as this in my own implementation of EDMM support for OE and it works (the EACCEPT successfully completes). 

Thank you all for the help.

**Comment by mingweishih** [comment]
> Can you explain why the hardcoded range of user_end-40*OE_PAGE_SIZE to user_end works? I tried setting it as this in my own implementation of EDMM support for OE and it works (the EACCEPT successfully completes).
> 
> Thank you all for the help.

There is a contract between host and enclave
https://github.com/mingweishih/openenclave/blob/b07dc610c34b4cd3bc050d3b5427177273974a4d/host/sgx/create.c#L69
https://github.com/mingweishih/openenclave/blob/b07dc610c34b4cd3bc050d3b5427177273974a4d/host/sgx/create.c#L512

**Comment by haitaohuang** [comment]
See [design doc here](https://github.com/intel/sgx-emm/blob/main/design_docs/SGX_EMM.md#allocate-deallocate-enclave-memory) EMM would do EACCEPT for all ranges you allocated with public API sgx_mm_alloc, either on #PF (COMMIT_ON_DEMAND)or immediately (COMMIT_NOW). [sgx_mm_init](https://github.com/intel/sgx-emm/blob/main/design_docs/SGX_EMM.md#support-for-emm-initialization) requires reserves a user range for the public API. The range should not overlap with those already commited and loaded by runtime before enclave initialization (EINIT).  Please note the size of the reserved user range need be big enough to accommodate your anticipated usage and EMM consumes a small portion of it for bookkeeping. So you need test a case with your maximal dynamic range expected or just reserve a large address range conservatively.
